scenes/cslib_menu.txt

```diff
@@ -105,165 +77,173 @@
 
-*label _soft_reset
-*comment (re-)calculate the visible options (after a page navigation)
-*temp option_slot_id_1 0
-*temp option_slot_id_2 0
-*temp option_slot_id_3 0
-*temp option_slot_id_4 0
-*temp option_slot_id_5 0
-*temp option_slot_id_6 0
-*temp option_slot_id_7 0
-*temp option_slot_id_8 0
-*temp option_slot_id_9 0
-*temp option_slot_id_10 0
+*label _build_array_common
+*params p_prefix p_cancel
+*temp prefix p_prefix
+*gosub _init_state
+*temp allow_cancel p_cancel
+*gosub _populate_page
+*return
 
-*temp ignore 0
-*set ignore ((cur_page - 1) * items_per_page)
-*temp item_count 0
-*temp cur_option 1
-*temp cur_item "@{invert_order ${{prefix&\"_len\"}}|1}"
-*label _filter_loop
-*if (filter = "")
-	*goto _do_not_filter
-*if ({((prefix&"_")&(filter&"_"))&cur_item} = filter_value)
-	*if (filter_type = "NOT_MATCH")
-		*goto _acc
-*elseif (filter_type != "NOT_MATCH")
-	*goto _acc
-*label _do_not_filter
-*set item_count + 1
-*if (ignore < 1) and (cur_option <= items_per_page)
-	*set option_slot_id[cur_option] cur_item
-	*set cur_option + 1
-*if (ignore > 0)
-	*set ignore -1
-*label _acc
-*if (invert_order)
-	*if (cur_item > 1)
-		*set cur_item - 1
-		*goto _filter_loop
-*else
-	*if (cur_item < {prefix&"_count"})
-		*set cur_item + 1
-		*goto _filter_loop
-*if (item_count = 0) and not(allow_cancel)
-	*bug cslib_menu: menu construction failed as there were 0 matching items — consider allowing cancellation
 
-*label _render
-*temp page_count 0
-*if (item_count < items_per_page)
-	*set page_count 1
-*else
-	*set page_count (item_count / items_per_page)
-*comment this *if might be redundant?
-*if (page_count < 1)
-	*set cslib_ret 0
-	*goto _return
+*label _init_state
+*comment determines the next/prev option position for smoother navigation
+*temp next_pref true
+*comment determines the direction in which options are displayed
+*temp const_invert_order false
+*comment the max amount of options to display per 'page'
+*temp const_items_per_page 5
+*if (const_items_per_page > 10)
+	*bug cslib_menu._init_state: const_items_per_page cannot be greater than 10
+*comment the text displayed on the cancel/return button
+*temp const_cancel_text "[return]"
+*temp page_count "@{({{prefix&\"_count\"}} <= const_items_per_page) 1|${({{prefix&\"_count\"}} / const_items_per_page)}}"
 *if ((page_count modulo 1) > 0)
 	*set page_count ((page_count - (page_count modulo 1)) + 1)
-*temp page_modifier (items_per_page * (cur_page - 1))
+*temp cur_page 1
+*return
+
+*label _populate_page
+*comment populate the choice items and check the guards
+*temp page_item_count 0
+*temp slot 1
+*gosub _init_option_vars
+*label _populate_page_loop
+*if (slot <= const_items_per_page)
+	*comment special handling for simple menus
+	*temp offset slot+(const_items_per_page * (cur_page -1))
+	*set option_guard[slot] ({prefix&"_count"} >= offset) and (const_items_per_page >= slot)
+	*if (option_guard[slot])
+		*set page_item_count + 1
+		*set option_text[slot] "${{prefix&\"_${offset}\"}}"
```

Comment:
would this work?
*set option_text[slot] {prefix & "_${offset}" }

also, if you change the prefix to include the _ (new *temp prefix_)  you can simply do

*set option_text[slot] {prefix_ & offset}

Reply:
Thanks for this, it's a really good suggestion.
So I've added the _ to the prefix variable, and also moved the arr_count out into an array. It's referenced a lot, and was indeed very ugly to look at.

Things look much better now:

```
*temp arr_count {{prefix&"count"}}
*temp page_count "@{(arr_count <= const_items_per_page) 1|${(arr_count / const_items_per_page)}}"
```